src/main/java/com/aws/greengrass/deployment/ConnectivityValidator.java

```diff
@@ -0,0 +1,235 @@
+        Topics newMqtt = deploymentConfiguration.getMQTTNamespace();
+        return !Topics.compareChildren(currentMqtt, newMqtt);
+    }
+
+    private boolean iotDataEndpointHasChanged(DeviceConfiguration currentdeviceConfiguration) {
+        // Defaults to empty string topic
+        Topic currentIotDataEndpoint = currentdeviceConfiguration.getIotDataEndpoint();
+        Topic newIotDataEndpoint = deploymentConfiguration.getIotDataEndpoint();
+        return !Topic.compareValue(currentIotDataEndpoint, newIotDataEndpoint);
+    }
+
+    private boolean greengrassDataPlaneEndpointHasChanged(DeviceConfiguration currentdeviceConfiguration) {
+        // Defaults to empty string topic
+        Topic currentGreengrassDataPlaneEndpoint = currentdeviceConfiguration.getGGDataEndpoint();
+        Topic newGreengrassDataPlaneEndpoint = deploymentConfiguration.getGGDataEndpoint();
+        return !Topic.compareValue(currentGreengrassDataPlaneEndpoint, newGreengrassDataPlaneEndpoint);
+    }
+
+    private boolean greengrassDataPlanePortHasChanged(DeviceConfiguration currentdeviceConfiguration) {
+        Topic currentGreengrassDataPlanePort = currentdeviceConfiguration.getGreengrassDataPlanePort();
+        Topic newGreengrassDataPlanePort = deploymentConfiguration.getGreengrassDataPlanePort();
+        return !Topic.compareValue(currentGreengrassDataPlanePort, newGreengrassDataPlanePort);
+    }
+
+    /**
+     * Creates an MQTT client and checks if the device can connect to AWS.
+     *
+     * @param totallyCompleteFuture Future will be updated if validation fails
+     */
+    public boolean mqttClientCanConnect(CompletableFuture<DeploymentResult> totallyCompleteFuture) {
+        logger.atDebug().log("Checking MQTT client can connect");
+
+        MqttClientConnection connection = null;
+        String message;
+        try (AwsIotMqttConnectionBuilder builder = mqttClient.createMqttConnectionBuilder(deploymentConfiguration,
+                securityService, null)) {
+            String clientId = Coerce.toString(deploymentConfiguration.getThingName()) + clientIdSuffix;
+            connection = builder.withClientId(clientId).build();
+            int operationTimeoutMillis = MqttClient.getMqttOperationTimeoutMillis(deploymentConfiguration);
+            connection.connect().get(operationTimeoutMillis, TimeUnit.MILLISECONDS);
+            return true;
+        } catch (MqttException e) {
+            message = "Mqtt client failed to connect: " + e.getMessage();
+            logger.atError().cause(e).log(message + withNewConfigMessageSuffix);
+        } catch (TimeoutException e) {
+            message = "Mqtt client validation timed out"; // exception has no message
+            logger.atError().cause(e).log(message + withNewConfigMessageSuffix);
+        } catch (ExecutionException e) {
+            message = "Mqtt client validation completed exceptionally: " + e.getMessage();
```

Comment:
ExecutionException is just a wrapper around the actual exception. You should first get the actual cause with `e.getCause()` and then check if there is a message in that exception which can be logged.

Reply:
It's already logged with `.cause(e)` is this something we also want to display on the console?

Reply:
we should be consistent about what we log on the device and the message we display on the console. so yes.